### ISSUE TITLE
Spurious problems when running in a multi-threaded environment

### ISSUE THREAD

**Issue by aunkrig** [janino-compiler/janino#32]
This problem was originally reported by Matthias Boehm from IBM:

> We recently ran into race conditions of multi-threaded
> compilation and loading of classes with the same class name. We fixed it
> by simply synchronizing the code path for code generation in Spark
> executors. However, it might be an indicator of unwanted side effects
> because the APIs of isolated SimpleCompilerinstances would suggest
> thread-safeness. The intermittent failures we got were puzzling as the
> related code works just fine:
> * Operand stack underrun at offset 91
> * No applicable constructor/method found for actual parameters
> * Binary numeric promotion not possible on types "java.lang.Object" and
> "double"

**Comment by aunkrig** [comment]
Hello Matthias,
just to make sure I get you right: There are multiple threads, each has its own SimpleCompiler object, then all the threads compile compilation units that declare a class with the same name.
In other words: Do you invoke "cook()" exactly once for each SimpleCompiler?
It would be extremely helpful if you could provide some example code that illustrates what leads to the problem.
Also: How exactly do you synchronize to work around the probem?

**Comment by mboehm7** [comment]
Thanks again @aunkrig for picking up this issue. To give a little more context, this issue only showed up due to a bug in SystemML (missing synchronization). We use code generation in local operations and distributed operations on Spark. For the distributed operations, we ship the java source code via task closures to the Spark executors and compile it there in order to ensure robustness, for example, with regard to different JDK/JRE versions on different nodes. Furthermore, we use a simple static cache to reuse compiled classes across tasks and executor threads. However, due to missing synchronization, there were race conditions, which led to multiple threads compiling and loading the same class. Each invocation uses a new instance of SimpleCompiler and calls cook exactly once.

Here is the [entry point of class compilation ](https://github.com/apache/systemml/blob/master/src/main/java/org/apache/sysml/runtime/codegen/CodegenUtils.java#L101), 
[the JANINO usage](https://github.com/apache/systemml/blob/master/src/main/java/org/apache/sysml/runtime/codegen/CodegenUtils.java#L180), and 
[an example usage](https://github.com/apache/systemml/blob/master/src/main/java/org/apache/sysml/runtime/instructions/spark/SpoofSPInstruction.java#L459)

**Comment by mboehm7** [comment]
It's not easy to create a reproducible case for this issue, but I now have a local test that allows to track down issue 1 (`Operand stack underrun at offset 91`), which fails with the following exception:

```
Caused by: org.codehaus.janino.JaninoRuntimeException: codegen.TMP14.genexec(double[] a, int ai, SideInput[] b, double[] scalars, double[] c, int len, int rix): Operand stack underrun at offset 91
        at org.codehaus.janino.CodeContext.flowAnalysis(CodeContext.java:501)
        at org.codehaus.janino.CodeContext.flowAnalysis(CodeContext.java:541)
        at org.codehaus.janino.CodeContext.flowAnalysis(CodeContext.java:313)
        at org.codehaus.janino.UnitCompiler.compile(UnitCompiler.java:2839)
        at org.codehaus.janino.UnitCompiler.compileDeclaredMethods(UnitCompiler.java:1262)
        at org.codehaus.janino.UnitCompiler.compileDeclaredMethods(UnitCompiler.java:1234)
        at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:538)
        at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:420)
        at org.codehaus.janino.UnitCompiler.access$400(UnitCompiler.java:206)
        at org.codehaus.janino.UnitCompiler$2.visitPackageMemberClassDeclaration(UnitCompiler.java:374)
        at org.codehaus.janino.UnitCompiler$2.visitPackageMemberClassDeclaration(UnitCompiler.java:369)
        at org.codehaus.janino.Java$AbstractPackageMemberClassDeclaration.accept(Java.java:1309)
        at org.codehaus.janino.UnitCompiler.compile(UnitCompiler.java:369)
        at org.codehaus.janino.UnitCompiler.compileUnit(UnitCompiler.java:345)
        at org.codehaus.janino.SimpleCompiler.compileToClassLoader(SimpleCompiler.java:396)
        at org.codehaus.janino.SimpleCompiler.cook(SimpleCompiler.java:205)
        at org.codehaus.janino.SimpleCompiler.cook(SimpleCompiler.java:196)
```

This test uses a simple fixed thread-pool in the number of virtual cores and compiles 1 million instances with
```

SimpleCompiler compiler = new SimpleCompiler();
compiler.cook(new String(TEMPLATE.getBytes()));
return compiler.getClassLoader()
	.loadClass("codegen.TMP14")
	.newInstance();
```

of the following class:

```
codegen.SpoofCompiler: package codegen;
import org.apache.sysml.runtime.codegen.LibSpoofPrimitives;
import org.apache.sysml.runtime.codegen.SpoofOperator.SideInput;
import org.apache.sysml.runtime.codegen.SpoofRowwise;
import org.apache.sysml.runtime.codegen.SpoofRowwise.RowType;
import org.apache.commons.math3.util.FastMath;

public final class TMP14 extends SpoofRowwise { 
  public TMP14() {
    super(RowType.ROW_AGG, -1, false, 0);
  }
  protected void genexec(double[] a, int ai, SideInput[] b, double[] scalars, double[] c, int len, int rix) { 
    double TMP0 = LibSpoofPrimitives.dotProduct(a, b[1].values(rix), ai, 0, len);
    double TMP1 = 1 / (1 + FastMath.exp(-TMP0));
    double TMP2 = getValue(b[0], rix);
    double TMP3 = (TMP1 < TMP2) ? 1 : 0;
    double TMP4 = 1 - 2 * TMP3;
    double TMP5 = TMP4 + 3;
    double TMP6 = TMP5 / 2;
    c[rix] = TMP6;
  }
  protected void genexec(double[] avals, int[] aix, int ai, SideInput[] b, double[] scalars, double[] c, int alen, int len, int rix) { 
    double TMP7 = LibSpoofPrimitives.dotProduct(avals, b[1].values(rix), aix, ai, 0, alen);
    double TMP8 = 1 / (1 + FastMath.exp(-TMP7));
    double TMP9 = getValue(b[0], rix);
    double TMP10 = (TMP8 < TMP9) ? 1 : 0;
    double TMP11 = 1 - 2 * TMP10;
    double TMP12 = TMP11 + 3;
    double TMP13 = TMP12 / 2;
    c[rix] = TMP13;
  }
}
``` 

To compile this you would need some SystemML classes. So, either you could compile SystemML via `mvn package` or I can send you snapshot jar if needed. Alternatively, it might give you an idea how to create an even smaller testcase.

Furthermore, I'd like to note that it might be dependent on the degree of parallelism or underlying platform. I ran it on a box with 2 sockets and E5-2440 (in total 24 virtual cores), CentOS Linux 7.4, and the following OpenJDK

```
openjdk version "1.8.0_151"
OpenJDK Runtime Environment (build 1.8.0_151-b12)
OpenJDK 64-Bit Server VM (build 25.151-b12, mixed mode)
```

**Comment by aunkrig** [comment]
I tried to reproduce the problem with your example code and as little changes as possible:

https://github.com/janino-compiler/janino/blob/master/commons-compiler-tests/src/test/java/org/codehaus/commons/compiler/tests/issue32/Issue32Test.java

But the problem didn't occur.
Can you pleas run my test case in _your_ Environment?

**Comment by aunkrig** [comment]
Tested with MS WINDOWS JREs 1.6, 1.7 and 1.8.

**Comment by aunkrig** [comment]
New idea:
If you set the logger org.codehaus.janino.CodeContext to level "FINE", then the "Operand stack underrun" error will only be _logged_ (and no exception is thrown), and the - probably inconsistent - class file is created and can be examined.
Ideally you could provide the inconsistent class file, and, for reference, also a consistent class file that was written by a _successful_ compilation. This way, it will be possible to see whether the _code generation_ is broken, or the following _flow analysis_.

**Comment by mboehm7** [comment]
after setting the log level to FINE, I got the following output
```
Nov 24, 2017 6:04:32 PM org.codehaus.janino.CodeContext flowAnalysis
FINE: codegen.TMP14.genexec(double[] a, int ai, SideInput[] b, double[] scalars, double[] c, int len, int rix): Operand stack underrun at offset 91
Nov 24, 2017 6:04:32 PM org.codehaus.janino.CodeContext flowAnalysis
FINE: genexec(double[] a, int ai, SideInput[] b, double[] scalars, double[] c, int len, int rix): Unexamined code at offset 92
```
and the resulting class is indeed corrupted which leads to the following exception on the subsequent `newInstance` call.

```
java.util.concurrent.ExecutionException: java.lang.VerifyError: (class: codegen/TMP14, method: genexec signature: ([DI[Lorg/apache/sysml/runtime/codegen/SpoofOperator$SideInput;[D[DII)V) Expecting to find double on stack
        at java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.util.concurrent.FutureTask.get(FutureTask.java:192)
        at org.apache.sysml.api.ThreadSafeTest.run(ThreadSafeTest.java:59)
        at org.apache.sysml.api.ThreadSafeTest.main(ThreadSafeTest.java:83)
Caused by: java.lang.VerifyError: (class: codegen/TMP14, method: genexec signature: ([DI[Lorg/apache/sysml/runtime/codegen/SpoofOperator$SideInput;[D[DII)V) Expecting to find double on stack
        at java.lang.Class.getDeclaredConstructors0(Native Method)
        at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
        at java.lang.Class.getConstructor0(Class.java:3075)
        at java.lang.Class.newInstance(Class.java:412)
        at org.apache.sysml.api.ThreadSafeTest$CompileTask.call(ThreadSafeTest.java:74)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by aunkrig** [comment]
Very good! Now, can you catch the corrupt class file and upload it here?

**Comment by mboehm7** [comment]
sure - I captured the byte representation of these classes from janino's `ByteArrayClassLoader`:

https://github.com/mboehm7/misc/blob/master/tmp/janino/TMP14.class
https://github.com/mboehm7/misc/blob/master/tmp/janino/TMP14corrupt.class

**Comment by aunkrig** [comment]
Brilliant!  And here we have it:

    public final class TMP14 extends org.apache.sysml.runtime.codegen.SpoofRowwise {
    
        protected void
        genexec(
            double[] p1,
            int p2,
            org.apache.sysml.runtime.codegen.SpoofOperator$SideInput[] p3,
            double[] p4,
            double[] p5,
            int p6,
            int p7
        ) {
    ...
    // c[rix] = TMP6;
            aload           [double[] p5]
            iload           [int p7]
            dload           [v13]
    +       invokestatic    Double.valueOf(double) => Double
            dastore
            return
        }

And the line marked with the "+" exists only in the corrupt class file!!
Obviously JANINO believes that it should convert the value of "TMP6" from `double` to `Double` before storing it in `c[rix]` (which has type `double`) - bad idea.

**Comment by aunkrig** [comment]
It seems that JANINO - namely the ```IClassLoader``` sometimes confuses the primitive types with their wrapper types, which is close to https://github.com/janino-compiler/janino/issues/7 .

However to fix this problem, we definitely need a test case that reproduces it _on my machine_ (MS WINDOWS 10, dual core). I've tried serious static code analysis, but could not spot any possible reason.

The unit test that I wrote (org.codehaus.commons.compiler.tests.issue32.Issue32Test) does not trigger the problem. Can you please try and improve it?


CU

Arno

**Comment by mboehm7** [comment]
thanks for investigating - I'm currently traveling but I'll make another attempt to create a more reproducible test next week.

**Comment by aunkrig** [comment]
Hi Matthias,

that would be great... a really nasty bug that I‘d love to fix.

CU Arno

**Comment by mboehm7** [comment]
@aunkrig just to let you know - so far I was not able to construct a test that reliably reproduces the issue so it could be included into a testsuite (which might indicate that it originates from a side effect w/ JIT). However, I will keep an eye on it and and try again whenever I see similar issues.

**Comment by johnwcowan** [comment]
@mboehm7 This may be absolutely irrelevant and probably is, but perhaps the problem has something to do with the case insensitivity of NTFS?  Trying to load double.class would normally fail on a Linux system, but on a Windows or MacOS system it would load Double.class instead.  This could be tested from Linux by setting up an NTFS file system

**Comment by mboehm7** [comment]
thanks for the note @johnwcowan - the code was indeed compiled on windows and executed on linux (and we faced similar case-sensitive issue in tests before) but as nicely shown by @aunkrig it seems to be related to incorrect conversions between native double types and Double objects. I was never able to reproduce similar issues, once we fixed the bug on our side.

Btw, we've forked SystemML into SystemDS and later merged it back to Apache SystemDS. We're still happily using Janino for Java codegen and we truly appreciate your efforts on this project. The related paper [1] also shows the comparison with javac in Figure 14.

[1] http://www.vldb.org/pvldb/vol11/p1755-boehm.pdf
